feature_filter in addition to feature_re, support for explain_prediction

lopuhin opened the pull request:
~~This is very much WIP, also missing sklearn tree ensemble explain_prediction support.~~

This branch brings two features:
- support ``feature_filter`` (a function) along with ``feature_re``, so now it's possible to exclude certain features
- support feature filtering for all explain_prediction functions

Also ``eli5.utils.mask`` now handles sparse "vectors".

kmike commented on eli5/_feature_names.py:
```diff
@@ -110,6 +117,20 @@ def filtered_by_re(self, feature_re):
 
+    def handle_filter(self, feature_filter, feature_re, x=None):
+        if feature_re is not None and feature_filter:
+            raise TypeError('pass either feature_filter or feature_re')
```
I think ValueError is more common for such cases (see e.g. http://stackoverflow.com/questions/256222/which-exception-should-i-raise-on-bad-illegal-argument-combinations-in-python?rq=1)

lopuhin replied:
Aha, thanks for the catch! I chose TypeError cause it's raised for missing/extra arguments, but ValueError really makes more sense, thanks for the link)

kmike commented:
Sorry, it seems I introduced a merge conflict in https://github.com/TeamHG-Memex/eli5/pull/133 :(

lopuhin commented:
>  Sorry, it seems I introduced a merge conflict in #133 :(

np, that was hardly avoidable :)

lopuhin commented:
@kmike this is finally ready for review - I implemented feature filtering in explain_prediction for sklearn trees and ensembles, and added more tests.

kmike commented on eli5/sklearn/explain_prediction.py:
```diff
@@ -107,7 +109,9 @@ def explain_prediction_linear_classifier(clf, doc,
                                          feature_names=None,
-                                         vectorized=False
+                                         feature_re=None,
+                                         feature_filter=None,
```
Could you please add feature_filter to the docstring of this function?

kmike commented on eli5/sklearn/explain_prediction.py:
```diff
@@ -194,6 +197,8 @@ def explain_prediction_linear_regressor(reg, doc,
                                         targets=None,
                                         feature_names=None,
+                                        feature_re=None,
+                                        feature_filter=None,
```
Could you please add these arguments to the docstring?

kmike commented on eli5/sklearn/explain_weights.py:
```diff
@@ -155,7 +155,8 @@ def explain_linear_classifier_weights(clf,
                                       coef_scale=None,
-                                      feature_re=None
+                                      feature_re=None,
+                                      feature_filter=None,
```
Here feature_filter is also not in the docstring

kmike commented on eli5/xgboost.py:
```diff
@@ -97,7 +99,10 @@ def explain_prediction_xgboost(
-        vectorized=False):
+        vectorized=False,
+        feature_re=None,
+        feature_filter=None,
```
I don't mind any argument order, but it'd be nice to have it consistent - e.g. in `explain_prediction_tree_regressor` `vectorized` is the last argument.

lopuhin replied:
Right, I like that feature_re and feature_filter come after feature_names :) Fixed it here too in 724c63a

kmike commented on tests/test_sklearn_explain_prediction.py:
```diff
@@ -303,3 +307,33 @@ def test_explain_tree_classifier_text(clf, newsgroups_train_big):
     format_as_all(res, clf)
+
+
+def _assert_tree_explain_prediction_single_target(clf, X, feature_names):
```
What do you think about moving this function to the top of the file, closer to other assert functions?

lopuhin replied:
Yep, a good call, thanks! Fixed in 724c63a

kmike commented on tests/test_sklearn_explain_weights.py:
```diff
@@ -270,15 +270,18 @@ def test_explain_linear_feature_re(newsgroups_train, vec):
 
-    res = explain_weights(clf, vec=vec, feature_re='^ath')
+    res = explain_weights(
+        clf, vec=vec, feature_re='^ath')
```
why wrap this line? :)

lopuhin replied:
Hah, that line suffered through some changes :) Thank for the catch! Fixed in 724c63a

kmike commented:
Looks great, thanks @lopuhin!

lopuhin commented:
Thanks for review @kmike ! I added feature_name to docstrings (linking to explain_weights or explain_predictions docs, as for other common parameters), and fixed other issues.

kmike commented on eli5/sklearn/explain_prediction.py:
```diff
@@ -293,9 +297,26 @@ def explain_prediction_tree_classifier(
+    passed through ``vec`` or not. By default it is False, meaning that
+    if ``vec`` is not None, ``vec.transform([doc])`` is passed to the
+    classifier. Set it to False if you're using ``vec`` to get feature names,
+    but ``doc`` is already vectorized.
```
See https://github.com/TeamHG-Memex/eli5/pull/134#discussion_r95250606 :P

lopuhin replied:
haha, right! :) Fixed in ea0c16a

kmike commented on eli5/sklearn/explain_prediction.py:
```diff
@@ -371,9 +398,26 @@ def explain_prediction_tree_regressor(
+    passed through ``vec`` or not. By default it is False, meaning that
+    if ``vec`` is not None, ``vec.transform([doc])`` is passed to the
+    classifier. Set it to False if you're using ``vec`` to get feature names,
+    but ``doc`` is already vectorized.
```
😛

lopuhin commented:
Oh, looks like I'm still missing some docstrings, let me double-check

lopuhin commented:
> Oh, looks like I'm still missing some docstrings, let me double-check

Yeah, I missed some, I think now all feature_filter docstrings are in place, hopefully.

kmike commented on eli5/sklearn/explain_prediction.py:
```diff
@@ -371,9 +398,26 @@ def explain_prediction_tree_regressor(
+    ``feature_re`` and ``feature_filter`` parameters.
+
+    ``vec`` is a vectorizer instance used to transform
+    raw features to the input of the classifier ``clf``
```
there is no ``clf`` argument

lopuhin replied:
Must be some GH bug, I bet the comment was for the correct line (this https://github.com/TeamHG-Memex/eli5/pull/132/files/ea0c16abdd1629890ed97dd352d623c20e33de98#diff-5a5dfe2e408016505008ee23671cb61fR411), but now it's show in the wrong place for me.
Fixed in dec5f20. Thanks for taking care of the docs, I didn't check them careful enough.

kmike commented on eli5/sklearn_crfsuite/explain_weights.py:
```diff
@@ -17,15 +17,21 @@ def explain_weights_sklearn_crfsuite(crf,
+    """ Explain sklearn_crfsuite.CRF weights.
+
+    See :func:`eli5.explain_weights` for description of
+    ``top``, ``feature_names``,
```
`target_names` and `targets` are also documented there, and there is no `feature_names` argument for explain_weights_sklearn_crfsuite

kmike commented on eli5/xgboost.py:
```diff
@@ -97,10 +105,28 @@ def explain_prediction_xgboost(
+    ``feature_re`` and ``feature_filter`` parameters.
+
+    ``vec`` is a vectorizer instance used to transform
+    raw features to the input of the classifier ``clf``
```
there is no ``clf`` argument
